components/raftstore/src/store/peer.rs

```diff
@@ -355,25 +363,55 @@ impl Peer {
     }
 
     /// Sets commit group to the peer.
-    pub fn init_commit_group(&mut self, state: &mut GlobalReplicationState) {
+    pub fn init_replication_mode(&mut self, state: &mut GlobalReplicationState) {
         debug!("init commit group"; "state" => ?state, "region_id" => self.region_id, "peer_id" => self.peer.id);
+        if !self.get_store().region().get_peers().is_empty() {
+            state.calculate_commit_group(self.get_store().region().get_peers());
+            self.raft_group
+                .raft
+                .assign_commit_groups(&state.group_buffer);
+        }
+        self.replication_sync = false;
         if state.status.get_mode() == ReplicationMode::Majority {
             self.raft_group.raft.enable_group_commit(false);
             self.replication_mode_version = 0;
+            self.dr_auto_sync_state = DrAutoSyncState::Async;
             return;
         }
-        self.raft_group.raft.enable_group_commit(true);
         self.replication_mode_version = state.status.get_dr_auto_sync().state_id;
-        if self.get_store().region().get_peers().is_empty() {
-            return;
-        }
-        if state.status.get_dr_auto_sync().get_state() == DrAutoSyncState::Async {
-            return;
+        let enable = state.status.get_dr_auto_sync().get_state() != DrAutoSyncState::Async;
+        self.raft_group.raft.enable_group_commit(enable);
+        self.dr_auto_sync_state = state.status.get_dr_auto_sync().get_state();
+    }
+
+    /// Updates replication mode.
+    pub fn switch_replication_mode(&mut self, state: &Mutex<GlobalReplicationState>) {
+        self.replication_sync = false;
+        let mut guard = state.lock().unwrap();
+        let enable_group_commit = if guard.status.get_mode() == ReplicationMode::Majority {
+            self.replication_mode_version = 0;
+            self.dr_auto_sync_state = DrAutoSyncState::Async;
+            false
+        } else {
+            self.dr_auto_sync_state = guard.status.get_dr_auto_sync().get_state();
+            self.replication_mode_version = guard.status.get_dr_auto_sync().state_id;
+            guard.status.get_dr_auto_sync().get_state() != DrAutoSyncState::Async
+        };
+        if enable_group_commit {
+            guard.calculate_commit_group(self.region().get_peers());
+            let ids = mem::replace(
+                &mut guard.group_buffer,
+                Vec::with_capacity(self.region().get_peers().len()),
+            );
```

Comment:
Sometimes `group_buffer` is reused to save allocation, sometimes it's not reused for reducing lock time.

Reply:
But `group_buffer` is calculated by the peers of a given region, we should not just reuse it globally(for every region).

Reply:
It's reused as a pure memory buffer, just as its name indicates. It's cleared every time it's reused. When calculate the group id, a lock has to be acquired, so a buffer can be shared globally.

Reply:
Got it! But I still have confused about the use of `mem::replace` here, and maybe we can remove `pub` for `group_buffer` and return a reference of it by `calculate_commit_group`, in case of people access `group_buffer` without calling `calculate_commit_group` first.